elastic/kibana
# [Lens] Reverse colors should not reverse palette picker previews

## Summary

Fixes #110290

This PR is an attempt to address the #110290 problem, while changing a bit the palette panel behaviour:
when revering a color stops configuration, this PR makes an automatic transition of the palette from the current state to a custom palette.

This means that the current saved palette configuration with predefined palettes and reverse option should be migrated to the new configuration.
Because the palette picker is now showing predefined palettes always with a direction, selecting one of these palettes resets the `reverse` flag.

<img width="321" alt="Screenshot 2021-08-30 at 12 56 01" src="https://user-images.githubusercontent.com/924948/131329911-d656a7d4-3e6e-43d7-b221-1f969be701ae.png">

* [x] On reverse transition to a custom palette
	* It would be nice to have the picker popover scroll to the current selection
* [x] On new palette selection from the palette picker resets the `reverse` flag
* [x] add a migration script to handle the new reverse state


## Testing

* Create a new datatable and with a predefined palette reverse the color stops
	* Check on the palette picker that all palettes have not been reversed and the current selection is now the `Custom Palette`
* I've prepared a dashboard with 4 datatables using conditional coloring with a predefined palette (2 reversed + 2 regular) before the fix/migration (rename the extension from `txt` to `ndjson` and import it)

  * Exported SO: [preDatatableFix.txt](https://github.com/elastic/kibana/files/7105135/preDatatableFix.txt)


### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Migration should correctly work on current user datatables/heatmaps | Medium | Medium | Unit tests will verify that migration will touch only specific datatables and heatmaps affected by the issue. |
